frontend/src/metabase/visualizations/echarts/cartesian/waterfall/option/index.ts

```diff
@@ -1,49 +1,274 @@
 import type { TimelineEventsModel } from "../../timeline-events/types";
-import { getAxes } from "./axis";
+import { getChartMeasurements } from "../../utils/layout";
+import { getTimelineEventsSeries } from "../../timeline-events/option";
+import { buildAxes } from "../../option/axis";
+
+type WaterfallSeriesOptions =
+  | RegisteredSeriesOption["line"]
+  | RegisteredSeriesOption["bar"]
+  | RegisteredSeriesOption["candlestick"];
+
+const DEFAULT_BAR_WIDTH = `60%`;
+
+// Ensures bars are not too wide when there are just a few
+const getBarWidthPercent = (barsCount: number) => 1 / (1.4 * barsCount);
+
+const getBarWidth = (
+  xAxisModel: XAxisModel,
+  chartMeasurements: ChartMeasurements,
+  settings: ComputedVisualizationSettings,
+) => {
+  if (
+    settings["graph.x_axis.scale"] !== "timeseries" ||
+    !xAxisModel.timeSeriesInterval
+  ) {
+    return DEFAULT_BAR_WIDTH;
+  }
+
+  let dataPointsCount = xAxisModel.timeSeriesInterval.lengthInIntervals + 1;
+  if (settings["waterfall.show_total"]) {
+    dataPointsCount += 1;
+  }
+  return Math.max(
+    5,
+    (chartMeasurements.bounds.right - chartMeasurements.bounds.left) *
+      getBarWidthPercent(dataPointsCount),
+  );
+};
+
+export const buildEChartsWaterfallSeries = (
+  chartModel: BaseCartesianChartModel,
+  settings: ComputedVisualizationSettings,
+  chartMeasurements: ChartMeasurements,
+  renderingContext: RenderingContext,
+): WaterfallSeriesOptions[] => {
+  const { seriesModels, transformedDataset: dataset } = chartModel;
+  const [seriesModel] = seriesModels;
+  const barWidth = getBarWidth(
+    chartModel.xAxisModel,
+    chartMeasurements,
+    settings,
+  );
+
+  const buildLabelOption = (key: DataKey) => ({
+    ...buildEChartsLabelOptions(
+      seriesModel,
+      settings,
+      renderingContext,
+      settings["graph.show_values"],
+    ),
+    formatter: getDataLabelFormatter(
+      seriesModel,
+      settings,
+      key,
+      renderingContext,
+      {
+        negativeInParentheses: true,
+      },
+    ),
+  });
+
+  const series: WaterfallSeriesOptions[] = [
+    {
+      id: seriesModel.dataKey,
+      type: "candlestick",
+      itemStyle: {
+        color: settings["waterfall.increase_color"],
+        color0: settings["waterfall.decrease_color"],
+        borderColor: "transparent",
+        borderColor0: "transparent",
+        borderColorDoji: "transparent",
+        borderWidth: 0,
+      },
+      animationDuration: 0,
+      barWidth,
+      dimensions: [
+        X_AXIS_DATA_KEY,
+        WATERFALL_START_KEY,
+        WATERFALL_END_KEY,
+        WATERFALL_START_2_KEY,
+        WATERFALL_END_2_KEY,
+      ],
+      encode: {
+        x: X_AXIS_DATA_KEY,
+        y: [
+          WATERFALL_START_KEY,
+          WATERFALL_END_KEY,
+          WATERFALL_START_2_KEY,
+          WATERFALL_END_2_KEY,
+        ],
+      },
+      zlevel: CHART_STYLE.series.zIndex,
+    },
+    {
+      id: "waterfall_labels",
```

Comment:
Let's extract this to a constant and add a prefix like the null character we used in `constants/dataset.ts` to avoid a potential collision with series names.